### ISSUE TITLE
[Feature Request] Better subqueries support

### ISSUE THREAD

**Issue by MaceWindu** [linq2db/linq2db#266]
Currently code like that:

```
...Select(_ => new Tuple<Guid, int>(_.Id, SqlExpressionFunction())).Where(_ => _.Item1 == id).Select(_ => _.Item2)
```

Fails trying to process Where() expression in SelectContext.ProcessScalar method on last line with NotImplementedException, because it supports only Parameter expression types for previous select, but we have New there. I understand that we cannot support any New expressions, because constructed types could have logic in constructors, but I think it's safe to support at least Tuple and anonymous types.

Or maybe you can propose how to do such query with linq2db:

```
SELECT t.rn
    FROM (
        SELECT TOP 1000000 id, ROW_NUMBER() OVER(ORDER BY (SELECT 0)) as rn
            FROM SomeTable
            ORDER BY name) as t
    WHERE t.id = @RecordId
```

Here I'm trying to load row number for specific record in ordered list. My problem is that I cannot use SQL query, because I need to have it as a generic helper applied to some base sorted query, so I don't know:
- FROM expression (could have joins)
- ORDER BY expression (could have multiple fields)
- also could have WHERE filters on base queries

Also there is another problem:
 I've tried to replace Tuple with anonymous type to test how it works with it and query executed, but only due to query optimization, linq2db generated completely different query with single select:

In C#:

```
query.Select(_ => new {
                _.Id,
                RowNumber = LinqExtensions.RowNumber()
            }).Take(int.MaxValue).Where(_ => _.Id == id).Select(_ => _.RowNumber).Single();
```

In SQL:

```
SELECT TOP (2)
    ROW_NUMBER() OVER(ORDER BY (SELECT 0)) as [c1]
FROM
    [Core$Designs] [t1]
WHERE
    [t1].[id] = @id
ORDER BY
    [t1].[creation_time] DESC
```

I think if I will find a way to tell linq2db my intentions it will fail with the same NotImplementedException.

**Comment by MaceWindu** [comment]
Managed to workaround it using extra join:
```
query
                .ApplySort(sort)
                .Select(_ => new
                {
                    Id = _.Id,
                    RowNumber = RowNumber()
                })
                .Take(int.MaxValue)
                .Join(query, _ => _.Id, _ => _.Id, (r, l) => new { r.RowNumber , l.Id })
                .Where(_ => _.Id.Equals(id))
                .Select(_ => _.RowNumber)
                .Single();
```

**Comment by MaceWindu** [comment]
Fixed by #1954
